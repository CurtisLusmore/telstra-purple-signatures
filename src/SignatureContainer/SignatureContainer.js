import React, { Component } from 'react';
import { entries } from 'lodash/object';

import constants from '../constants';
import Signature from '../Signature';
import RepliesAndForwards from '../RepliesAndForwards';
import BtsRepliesAndForwards from '../BtsRepliesAndForwards';
import BtsSignature from '../BtsSignature';
import {
  copySignatureText,
  copyRepliesAndForwardsText,
  copySignature,
  copyBtsSignatureText,
  copyBtsRepliesAndForwardsText
} from '../util';
import Button from '../Button';

import './SignatureContainer.scss';

const ReadifySignatureContainer = ({
  signatureProps,
  CopySignatureText,
  CopySignatureHtml,
  CopyRepliesAndForwardsText,
  CopyRepliesAndForwardsHtml
}) => (
  <div className="content">
    <div className="level">
      <div className="level-left">
        <h3 className="level-item">Standard Signature</h3>
      </div>
      <div className="level-right">
        {CopySignatureText}
        {CopySignatureHtml}
      </div>
    </div>
    <Signature {...signatureProps} />
    <div className="level">
      <div className="level-left">
        <h3 className="level-item">Replies and Forwards Signature</h3>
      </div>
      <div className="level-right">
        {CopyRepliesAndForwardsText}
        {CopyRepliesAndForwardsHtml}
      </div>
    </div>
    <RepliesAndForwards {...signatureProps} />
  </div>
);

const BtsSignatureContainer = ({
  btsProps,
  CopyBtsSigText,
  CopyBtsHtml,
  CopyBtsRepliesAndForwardsText,
  CopyBtsRepliesAndForwardsHtml
}) => (
  <div className="content">
    <div className="level">
      <div className="level-left">
        <h3 className="level-item">Standard Signature</h3>
      </div>
      <div className="level-right">
        {CopyBtsSigText}
        {CopyBtsHtml}
      </div>
    </div>
    <BtsSignature {...btsProps} />
    <div className="level">
      <div className="level-left">
        <h3 className="level-item">Replies and Forwards Signature</h3>
      </div>
      <div className="level-right">
        {CopyBtsRepliesAndForwardsText}
        {CopyBtsRepliesAndForwardsHtml}
      </div>
    </div>
    <BtsRepliesAndForwards {...btsProps} />
  </div>
);

class SignatureContainer extends Component {
  static defaultButtonText = {
    text: 'Copy text only',
    html: 'Copy signature'
  };

  static buttonClasses = {
    text: 'button',
    html: 'button is-primary'
  };

  buttonMaker = (clickHandler, isText) => {
    const isSuccessClass = 'button is-success';
    const copiedText = 'Copied!';
    const {
      text: classText,
      html: classHtml
    } = SignatureContainer.buttonClasses;
    const {
      text: textText,
      html: textHtml
    } = SignatureContainer.defaultButtonText;
    return (
      <Button
        onClickHandler={clickHandler}
        classBefore={isText ? classText : classHtml}
        classAfter={isSuccessClass}
        textBefore={isText ? textText : textHtml}
        textAfter={copiedText}
      />
    );
  };

  createButtons = (signatureProps, btsProps, placeholders) => {
    const CopySignatureText = this.buttonMaker(
      () =>
        copySignatureText({
          ...signatureProps,
          placeholders: placeholders
        }),
      true
    );
    const CopySignatureHtml = this.buttonMaker(
      () =>
        copySignature(
          {
            ...signatureProps,
            placeholders: placeholders
          },
          Signature
        ),
      false
    );
    const CopyRepliesAndForwardsText = this.buttonMaker(
      () =>
        copyRepliesAndForwardsText({
          ...signatureProps,
          placeholders: placeholders
        }),
      true
    );
    const CopyRepliesAndForwardsHtml = this.buttonMaker(
      () =>
        copySignature(
          {
            ...signatureProps,
            placeholders: placeholders
          },
          RepliesAndForwards
        ),
      false
    );

    const CopyBtsSigText = this.buttonMaker(
      () =>
        copyBtsSignatureText({
          ...btsProps,
          placeholders: placeholders
        }),
      true
    );
    const CopyBtsHtml = this.buttonMaker(
      () =>
        copySignature(
          {
            ...btsProps,
            placeholders: placeholders
          },
          BtsSignature
        ),
      false
    );

    const CopyBtsRepliesAndForwardsText = this.buttonMaker(
      () =>
        copyBtsRepliesAndForwardsText({
          ...btsProps,
          placeholders: placeholders
        }),
      true
    );
    const CopyBtsRepliesAndForwardsHtml = this.buttonMaker(
      () =>
        copySignature(
          {
            ...btsProps,
            placeholders: placeholders
          },
          BtsRepliesAndForwards
        ),
      false
    );
    return {
      CopySignatureText,
      CopySignatureHtml,
      CopyRepliesAndForwardsText,
      CopyRepliesAndForwardsHtml,
      CopyBtsSigText,
      CopyBtsHtml,
      CopyBtsRepliesAndForwardsText,
      CopyBtsRepliesAndForwardsHtml
    };
  };

  assignPlaceholders = (props, placeholders) => {
    const isBlank = a => a === '' || a === null || a === undefined;
    return entries(props).reduce((result, item) => {
      const [key, value] = item;
      return Object.assign(result, {
        [key]: isBlank(value) ? placeholders[key] : value
      });
    }, {});
  };

  render() {
    const {
      name,
      title,
      qualifications,
      twitter,
      sigType,
      mobile,
      email,
      phone
    } = this.props;

    const placeholders =
      constants[sigType === 'bts' ? 'btsDigital' : 'readify'].placeholders;

    const signatureProps = this.assignPlaceholders(
      {
        name,
        title,
        qualifications,
        mobile,
        email,
        twitter,
        isSupport: sigType === 'support'
      },
      placeholders
    );

    const btsProps = this.assignPlaceholders(
      {
        name,
        title,
        qualifications,
        mobile,
        phone,
        email
      },
      placeholders
    );

    const {
      CopySignatureText,
      CopySignatureHtml,
      CopyRepliesAndForwardsText,
      CopyRepliesAndForwardsHtml,
      CopyBtsSigText,
      CopyBtsHtml,
      CopyBtsRepliesAndForwardsText,
      CopyBtsRepliesAndForwardsHtml
    } = this.createButtons(signatureProps, btsProps, placeholders);

    const readifySignatureContainerProps = {
      signatureProps,
      CopySignatureText,
      CopySignatureHtml,
      CopyRepliesAndForwardsText,
      CopyRepliesAndForwardsHtml
    };
    const BtsSignatureContainerProps = {
      btsProps,
      CopyBtsSigText,
      CopyBtsHtml,
      CopyBtsRepliesAndForwardsText,
      CopyBtsRepliesAndForwardsHtml
    };

    return sigType === 'bts' ? (
      <BtsSignatureContainer {...BtsSignatureContainerProps} />
    ) : (
      <ReadifySignatureContainer {...readifySignatureContainerProps} />
    );
  }
}

export default SignatureContainer;
