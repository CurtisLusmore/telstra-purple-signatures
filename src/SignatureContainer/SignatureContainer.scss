.content {
  table {
    tr {
      td {
        padding-left: 0;
        border: none;
      }
    }
  }
}