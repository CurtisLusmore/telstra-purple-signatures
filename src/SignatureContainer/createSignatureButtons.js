import React from 'react';

import {
  copyBtsRepliesAndForwardsText,
  copyBtsSignatureText,
  copyRepliesAndForwardsText,
  copySignature,
  copySignatureText
} from '../util';
import Signature from '../Signature';
import RepliesAndForwards from '../RepliesAndForwards';
import BtsSignature from '../BtsSignature';
import BtsRepliesAndForwards from '../BtsRepliesAndForwards';
import Button from '../Button';

const buttonMaker = (clickHandler, isText) => {
  const isSuccessClass = 'button is-success';
  const copiedText = 'Copied!';
  const classText = 'button';
  const classHtml = 'button is-primary';
  const textText = 'Copy text only';
  const textHtml = 'Copy signature';
  return (
    <Button
      onClickHandler={clickHandler}
      classBefore={isText ? classText : classHtml}
      classAfter={isSuccessClass}
      textBefore={isText ? textText : textHtml}
      textAfter={copiedText}
    />
  );
};
export const createButtons = (signatureProps, placeholders, isBts) => {
  const clickHandlers = {
    CopySignatureText: () =>
      copySignatureText({
        ...signatureProps,
        placeholders: placeholders
      }),
    CopySignatureHtml: () =>
      copySignature(
        {
          ...signatureProps,
          placeholders: placeholders
        },
        Signature
      ),
    CopyRepliesAndForwardsText: () =>
      copyRepliesAndForwardsText({
        ...signatureProps,
        placeholders: placeholders
      }),
    CopyRepliesAndForwardsHtml: () =>
      copySignature(
        {
          ...signatureProps,
          placeholders: placeholders
        },
        RepliesAndForwards
      ),
    CopyBtsSigText: () =>
      copyBtsSignatureText({
        ...signatureProps,
        placeholders: placeholders
      }),
    CopyBtsHtml: () =>
      copySignature(
        {
          ...signatureProps,
          placeholders: placeholders
        },
        BtsSignature
      ),
    CopyBtsRepliesAndForwardsText: () =>
      copyBtsRepliesAndForwardsText({
        ...signatureProps,
        placeholders: placeholders
      }),
    CopyBtsRepliesAndForwardsHtml: () =>
      copySignature(
        {
          ...signatureProps,
          placeholders: placeholders
        },
        BtsRepliesAndForwards
      )
  };

  return [
    { name: 'CopySignatureText', isText: true, bts: false },
    { name: 'CopySignatureHtml', isText: false, bts: false },
    { name: 'CopyRepliesAndForwardsText', isText: true, bts: false },
    { name: 'CopyRepliesAndForwardsHtml', isText: false, bts: false },
    { name: 'CopyBtsSigText', isText: true, bts: true },
    { name: 'CopyBtsHtml', isText: false, bts: true },
    { name: 'CopyBtsRepliesAndForwardsText', isText: true, bts: true },
    { name: 'CopyBtsRepliesAndForwardsHtml', isText: false, bts: true }
  ]
    .filter(objBase => objBase.bts === isBts)
    .map(buttonArr => [
      buttonArr.name,
      buttonMaker(clickHandlers[buttonArr.name], buttonArr.isText)
    ])
    .reduce(
      (result, item) => Object.assign(result, { [item[0]]: item[1] }),
      {}
    );
};
