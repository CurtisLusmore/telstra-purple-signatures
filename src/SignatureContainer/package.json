{
  "main": "./SignatureContainer.js",
  "name": "SignatureContainer",
  "version": "0.0.0",
  "private": true
}