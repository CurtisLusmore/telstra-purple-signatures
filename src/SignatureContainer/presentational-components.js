import React from 'react';

import Signature from '../Signature';
import RepliesAndForwards from '../RepliesAndForwards';
import BtsSignature from '../BtsSignature';
import BtsRepliesAndForwards from '../BtsRepliesAndForwards';

export const ReadifySignatureContainer = ({
  signatureProps,
  CopySignatureText,
  CopySignatureHtml,
  CopyRepliesAndForwardsText,
  CopyRepliesAndForwardsHtml
}) => (
  <div className="content">
    <div className="level">
      <div className="level-left">
        <h3 className="level-item">Standard Signature</h3>
      </div>
      <div className="level-right">
        {CopySignatureText}
        {CopySignatureHtml}
      </div>
    </div>
    <Signature {...signatureProps} />
    <div className="level">
      <div className="level-left">
        <h3 className="level-item">Replies and Forwards Signature</h3>
      </div>
      <div className="level-right">
        {CopyRepliesAndForwardsText}
        {CopyRepliesAndForwardsHtml}
      </div>
    </div>
    <RepliesAndForwards {...signatureProps} />
  </div>
);
export const BtsSignatureContainer = ({
  signatureProps,
  CopyBtsSigText,
  CopyBtsHtml,
  CopyBtsRepliesAndForwardsText,
  CopyBtsRepliesAndForwardsHtml
}) => (
  <div className="content">
    <div className="level">
      <div className="level-left">
        <h3 className="level-item">Standard Signature</h3>
      </div>
      <div className="level-right">
        {CopyBtsSigText}
        {CopyBtsHtml}
      </div>
    </div>
    <BtsSignature {...signatureProps} />
    <div className="level">
      <div className="level-left">
        <h3 className="level-item">Replies and Forwards Signature</h3>
      </div>
      <div className="level-right">
        {CopyBtsRepliesAndForwardsText}
        {CopyBtsRepliesAndForwardsHtml}
      </div>
    </div>
    <BtsRepliesAndForwards {...signatureProps} />
  </div>
);
