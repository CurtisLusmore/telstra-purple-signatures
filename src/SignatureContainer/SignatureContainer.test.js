import React from 'react';
import { shallow } from 'enzyme';

import SignatureContainer from './SignatureContainer';

it('Renders', () => {
  const props = {
    email: 'your.name@readify.net',
    sigType: 'readify',
    mobile: '+61 111 222 333',
    name: 'your name',
    title: 'Job Title',
    twitter: '@myTwitter'
  };

  const tree = shallow(<SignatureContainer {...props} />);
  expect(tree).toMatchSnapshot();
});

it('Renders with support fields', () => {
  const props = {
    email: 'your.name@readify.net',
    sigType: 'support',
    mobile: '+61 111 222 333',
    name: 'your name',
    title: 'Job Title',
    twitter: '@myTwitter'
  };

  const tree = shallow(<SignatureContainer {...props} />);
  expect(tree).toMatchSnapshot();
});
