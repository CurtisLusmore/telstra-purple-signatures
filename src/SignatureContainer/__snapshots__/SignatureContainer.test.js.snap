// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignatureContainer Renders 1`] = `
<mapProps(SignatureContainer)
  Container={[Function]}
  email="your.name@telstra.purple.net"
  mobile="+61 111 222 333"
  name="your name"
  sigType="purple"
  title="Job Title"
  twitter="@myTwitter"
/>
`;

exports[`SignatureContainer Renders with support fields 1`] = `
<mapProps(SignatureContainer)
  Container={[Function]}
  email="your.name@telstra.purple.net"
  mobile="+61 111 222 333"
  name="your name"
  sigType="support"
  title="Job Title"
  twitter="@myTwitter"
/>
`;

exports[`createContainer returns the right props 1`] = `
Object {
  "Container": undefined,
  "containerProps": Object {
    "CopyRepliesAndForwardsHtml": <withState(withState(withHandlers(mapProps(Button))))
      classAfter="button is-success"
      classBefore="button is-primary"
      onClickHandler={[Function]}
      textAfter="Copied!"
      textBefore="Copy signature"
    />,
    "CopyRepliesAndForwardsText": <withState(withState(withHandlers(mapProps(Button))))
      classAfter="button is-success"
      classBefore="button"
      onClickHandler={[Function]}
      textAfter="Copied!"
      textBefore="Copy text only"
    />,
    "CopySignatureHtml": <withState(withState(withHandlers(mapProps(Button))))
      classAfter="button is-success"
      classBefore="button is-primary"
      onClickHandler={[Function]}
      textAfter="Copied!"
      textBefore="Copy signature"
    />,
    "CopySignatureText": <withState(withState(withHandlers(mapProps(Button))))
      classAfter="button is-success"
      classBefore="button"
      onClickHandler={[Function]}
      textAfter="Copied!"
      textBefore="Copy text only"
    />,
    "signatureProps": Object {
      "email": "your.name@telstra.purple.net",
      "isSupport": false,
      "mobile": "+61 111 222 333",
      "name": "your name",
      "title": "Job Title",
      "twitter": "@myTwitter",
    },
  },
}
`;

exports[`createContainer returns the right props 2`] = `
Object {
  "Container": undefined,
  "containerProps": Object {
    "CopyRepliesAndForwardsHtml": <withState(withState(withHandlers(mapProps(Button))))
      classAfter="button is-success"
      classBefore="button is-primary"
      onClickHandler={[Function]}
      textAfter="Copied!"
      textBefore="Copy signature"
    />,
    "CopyRepliesAndForwardsText": <withState(withState(withHandlers(mapProps(Button))))
      classAfter="button is-success"
      classBefore="button"
      onClickHandler={[Function]}
      textAfter="Copied!"
      textBefore="Copy text only"
    />,
    "CopySignatureHtml": <withState(withState(withHandlers(mapProps(Button))))
      classAfter="button is-success"
      classBefore="button is-primary"
      onClickHandler={[Function]}
      textAfter="Copied!"
      textBefore="Copy signature"
    />,
    "CopySignatureText": <withState(withState(withHandlers(mapProps(Button))))
      classAfter="button is-success"
      classBefore="button"
      onClickHandler={[Function]}
      textAfter="Copied!"
      textBefore="Copy text only"
    />,
    "signatureProps": Object {
      "email": "your.name@telstra.purple.net",
      "isSupport": false,
      "mobile": "+61 111 222 333",
      "name": "your name",
      "title": "Job Title",
      "twitter": "@myTwitter",
    },
  },
}
`;
